File: 测试高级搜索功能.js
// 测试高级搜索和筛选功能
const API = require('./utils/api-manager.js').API

// 测试配置
const testConfig = {
  baseUrl: 'http://localhost:3000',
  timeout: 10000
}

// 测试用例
const testCases = [
  {
    name: '市场高级搜索',
    type: 'markets',
    filters: {
      keyword: '茶叶',
      province: '福建省',
      dateRange: '2024-01-01-2024-12-31'
    }
  },
  {
    name: '新品高级搜索',
    type: 'newarrivals',
    filters: {
      keyword: '龙井',
      category: '绿茶',
      priceRange: '100-500',
      dateRange: '2024-01-01-2024-12-31'
    }
  },
  {
    name: '供求高级搜索',
    type: 'supplies',
    filters: {
      keyword: '铁观音',
      type: '供应',
      priceRange: '200-1000',
      status: 'active'
    }
  },
  {
    name: '清仓高级搜索',
    type: 'clearance',
    filters: {
      keyword: '普洱',
      category: '普洱茶',
      priceRange: '50-300'
    }
  },
  {
    name: '内容高级搜索',
    type: 'content',
    filters: {
      keyword: '茶艺',
      type: 'art',
      tag: '冲泡技巧'
    }
  }
]

// 测试函数
async function testAdvancedSearch() {
  console.log('🚀 [测试] 开始测试高级搜索功能...')
  
  // 1. 测试连接
  console.log('\n📡 [测试] 检查服务器连接...')
  try {
    const healthResponse = await API.health()
    console.log('✅ 服务器连接正常:', healthResponse)
  } catch (error) {
    console.error('❌ 服务器连接失败:', error.message)
    return
  }
  
  // 2. 测试筛选选项获取
  console.log('\n🔧 [测试] 获取筛选选项...')
  try {
    const optionsResponse = await API.getFilterOptions('all')
    console.log('✅ 筛选选项获取成功:', optionsResponse.data)
  } catch (error) {
    console.error('❌ 筛选选项获取失败:', error.message)
  }
  
  // 3. 测试各种高级搜索
  for (const testCase of testCases) {
    console.log(`\n🔍 [测试] ${testCase.name}...`)
    
    try {
      let response
      const filters = API.buildAdvancedFilters(testCase.filters)
      
      switch (testCase.type) {
        case 'markets':
          response = await API.advancedSearchMarkets(filters)
          break
        case 'newarrivals':
          response = await API.advancedSearchNewarrivals(filters)
          break
        case 'supplies':
          response = await API.advancedSearchSupplies(filters)
          break
        case 'clearance':
          response = await API.advancedSearchClearance(filters)
          break
        case 'content':
          response = await API.advancedSearchContent(filters)
          break
        default:
          throw new Error('不支持的数据类型')
      }
      
      if (response.status === 'success') {
        console.log(`✅ ${testCase.name} 搜索成功:`)
        console.log(`   - 结果数量: ${response.data.length}`)
        console.log(`   - 总数量: ${response.pagination?.total || '未知'}`)
        console.log(`   - 筛选条件: ${JSON.stringify(testCase.filters)}`)
        
        // 显示前3条结果
        if (response.data.length > 0) {
          console.log(`   - 前3条结果:`)
          response.data.slice(0, 3).forEach((item, index) => {
            console.log(`     ${index + 1}. ${item.name || item.title || item.id}`)
          })
        }
      } else {
        console.error(`❌ ${testCase.name} 搜索失败:`, response.message)
      }
    } catch (error) {
      console.error(`❌ ${testCase.name} 搜索异常:`, error.message)
    }
  }
  
  // 4. 测试边界情况
  console.log('\n🔬 [测试] 边界情况测试...')
  
  // 空搜索
  try {
    const emptyResponse = await API.advancedSearchMarkets({})
    console.log('✅ 空搜索测试通过:', emptyResponse.data.length, '条结果')
  } catch (error) {
    console.error('❌ 空搜索测试失败:', error.message)
  }
  
  // 无效价格范围
  try {
    const invalidPriceResponse = await API.advancedSearchNewarrivals({
      priceRange: 'invalid-price'
    })
    console.log('✅ 无效价格范围处理正常:', invalidPriceResponse.data.length, '条结果')
  } catch (error) {
    console.error('❌ 无效价格范围处理失败:', error.message)
  }
  
  // 无效日期范围
  try {
    const invalidDateResponse = await API.advancedSearchSupplies({
      dateRange: 'invalid-date'
    })
    console.log('✅ 无效日期范围处理正常:', invalidDateResponse.data.length, '条结果')
  } catch (error) {
    console.error('❌ 无效日期范围处理失败:', error.message)
  }
  
  // 5. 测试分页功能
  console.log('\n📄 [测试] 分页功能测试...')
  try {
    const pageResponse = await API.advancedSearchMarkets({
      page: 1,
      perPage: 5
    })
    console.log('✅ 分页功能正常:')
    console.log(`   - 当前页: ${pageResponse.pagination.page}`)
    console.log(`   - 每页数量: ${pageResponse.pagination.per_page}`)
    console.log(`   - 总页数: ${pageResponse.pagination.pages}`)
    console.log(`   - 总数量: ${pageResponse.pagination.total}`)
  } catch (error) {
    console.error('❌ 分页功能测试失败:', error.message)
  }
  
  // 6. 测试组合搜索
  console.log('\n🎯 [测试] 组合搜索测试...')
  try {
    const complexFilters = {
      keyword: '茶',
      category: '绿茶',
      priceRange: '50-1000',
      dateRange: '2024-01-01-2024-12-31',
      page: 1,
      perPage: 10
    }
    
    const complexResponse = await API.advancedSearchNewarrivals(complexFilters)
    console.log('✅ 组合搜索测试通过:')
    console.log(`   - 筛选条件: ${JSON.stringify(complexFilters)}`)
    console.log(`   - 结果数量: ${complexResponse.data.length}`)
    console.log(`   - 总数量: ${complexResponse.pagination.total}`)
  } catch (error) {
    console.error('❌ 组合搜索测试失败:', error.message)
  }
  
  console.log('\n🎉 [测试] 高级搜索功能测试完成!')
}

// 测试参数构建功能
function testBuildAdvancedFilters() {
  console.log('\n🔧 [测试] 参数构建功能测试...')
  
  const testOptions = {
    keyword: '龙井茶',
    category: '绿茶',
    priceRange: '100-500',
    dateRange: '2024-01-01-2024-12-31',
    province: '浙江省',
    type: '供应',
    tag: '新品推荐',
    status: 'active',
    page: 1,
    perPage: 20
  }
  
  const filters = API.buildAdvancedFilters(testOptions)
  console.log('✅ 参数构建结果:', filters)
  
  // 验证参数
  const expectedKeys = [
    'keyword', 'category', 'price_range', 'date_range', 
    'province', 'type', 'tag', 'status', 'page', 'per_page'
  ]
  
  const missingKeys = expectedKeys.filter(key => !(key in filters))
  if (missingKeys.length === 0) {
    console.log('✅ 所有参数都正确构建')
  } else {
    console.error('❌ 缺少参数:', missingKeys)
  }
}

// 测试筛选选项功能
async function testFilterOptions() {
  console.log('\n📋 [测试] 筛选选项功能测试...')
  
  const dataTypes = ['markets', 'newarrivals', 'supplies', 'clearance', 'content', 'all']
  
  for (const dataType of dataTypes) {
    try {
      const response = await API.getFilterOptions(dataType)
      console.log(`✅ ${dataType} 筛选选项:`, response.data)
    } catch (error) {
      console.error(`❌ ${dataType} 筛选选项获取失败:`, error.message)
    }
  }
}

// 性能测试
async function performanceTest() {
  console.log('\n⚡ [测试] 性能测试...')
  
  const startTime = Date.now()
  
  try {
    // 并发搜索测试
    const promises = [
      API.advancedSearchMarkets({ keyword: '茶' }),
      API.advancedSearchNewarrivals({ category: '绿茶' }),
      API.advancedSearchSupplies({ type: '供应' }),
      API.advancedSearchClearance({ category: '普洱茶' }),
      API.advancedSearchContent({ type: 'recommend' })
    ]
    
    const results = await Promise.all(promises)
    const endTime = Date.now()
    const duration = endTime - startTime
    
    console.log(`✅ 并发搜索测试完成，耗时: ${duration}ms`)
    console.log(`   - 平均响应时间: ${duration / results.length}ms`)
    
    results.forEach((result, index) => {
      console.log(`   - 搜索${index + 1}: ${result.data.length} 条结果`)
    })
    
  } catch (error) {
    console.error('❌ 性能测试失败:', error.message)
  }
}

// 主测试函数
async function runAllTests() {
  console.log('🚀 [高级搜索测试] 开始全面测试...')
  console.log('=' * 50)
  
  try {
    // 基础功能测试
    await testAdvancedSearch()
    
    // 参数构建测试
    testBuildAdvancedFilters()
    
    // 筛选选项测试
    await testFilterOptions()
    
    // 性能测试
    await performanceTest()
    
    console.log('\n🎉 [测试完成] 所有测试通过!')
    console.log('=' * 50)
    
  } catch (error) {
    console.error('\n❌ [测试失败] 测试过程中出现错误:', error)
  }
}

// 导出测试函数
module.exports = {
  testAdvancedSearch,
  testBuildAdvancedFilters,
  testFilterOptions,
  performanceTest,
  runAllTests
}

// 如果直接运行此文件，执行所有测试
if (typeof module !== 'undefined' && module.exports) {
  // 在Node.js环境中运行
  runAllTests()
} else {
  // 在小程序环境中运行
  console.log('🔧 [高级搜索测试] 在小程序环境中运行测试...')
  runAllTests()
} 

File: 测试高级搜索功能.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
高级搜索和筛选功能测试脚本
"""

import requests
import json
import time
from datetime import datetime

# 测试配置
BASE_URL = 'http://localhost:3000'
TIMEOUT = 10

class AdvancedSearchTester:
    def __init__(self):
        self.base_url = BASE_URL
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'AdvancedSearchTester/1.0'
        })
    
    def test_health(self):
        """测试服务器健康状态"""
        print("📡 [测试] 检查服务器连接...")
        try:
            response = self.session.get(f"{self.base_url}/health", timeout=TIMEOUT)
            if response.status_code == 200:
                data = response.json()
                print(f"✅ 服务器连接正常: {data}")
                return True
            else:
                print(f"❌ 服务器连接失败: HTTP {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ 服务器连接异常: {e}")
            return False
    
    def test_filter_options(self):
        """测试筛选选项获取"""
        print("\n🔧 [测试] 获取筛选选项...")
        try:
            response = self.session.get(f"{self.base_url}/api/filter-options?type=all", timeout=TIMEOUT)
            if response.status_code == 200:
                data = response.json()
                print(f"✅ 筛选选项获取成功:")
                for key, value in data.get('data', {}).items():
                    print(f"   - {key}: {value}")
                return True
            else:
                print(f"❌ 筛选选项获取失败: HTTP {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ 筛选选项获取异常: {e}")
            return False
    
    def test_advanced_search_markets(self):
        """测试市场高级搜索"""
        print("\n🔍 [测试] 市场高级搜索...")
        test_cases = [
            {
                'name': '关键词搜索',
                'params': {'keyword': '茶叶'}
            },
            {
                'name': '省份筛选',
                'params': {'province': '福建省'}
            },
            {
                'name': '组合搜索',
                'params': {'keyword': '茶叶', 'province': '福建省'}
            },
            {
                'name': '分页测试',
                'params': {'page': 1, 'per_page': 5}
            }
        ]
        
        for test_case in test_cases:
            try:
                print(f"   - 测试: {test_case['name']}")
                response = self.session.get(
                    f"{self.base_url}/api/advanced-search/markets",
                    params=test_case['params'],
                    timeout=TIMEOUT
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data.get('status') == 'success':
                        print(f"     ✅ 成功: {len(data['data'])} 条结果")
                        if data['data']:
                            print(f"     📋 示例: {data['data'][0].get('name', 'N/A')}")
                    else:
                        print(f"     ❌ 失败: {data.get('message', '未知错误')}")
                else:
                    print(f"     ❌ HTTP错误: {response.status_code}")
                    
            except Exception as e:
                print(f"     ❌ 异常: {e}")
    
    def test_advanced_search_newarrivals(self):
        """测试新品高级搜索"""
        print("\n🔍 [测试] 新品高级搜索...")
        test_cases = [
            {
                'name': '关键词搜索',
                'params': {'keyword': '龙井'}
            },
            {
                'name': '分类筛选',
                'params': {'category': '绿茶'}
            },
            {
                'name': '价格范围',
                'params': {'price_range': '100-500'}
            },
            {
                'name': '组合搜索',
                'params': {'keyword': '龙井', 'category': '绿茶', 'price_range': '100-500'}
            }
        ]
        
        for test_case in test_cases:
            try:
                print(f"   - 测试: {test_case['name']}")
                response = self.session.get(
                    f"{self.base_url}/api/advanced-search/newarrivals",
                    params=test_case['params'],
                    timeout=TIMEOUT
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data.get('status') == 'success':
                        print(f"     ✅ 成功: {len(data['data'])} 条结果")
                        if data['data']:
                            print(f"     📋 示例: {data['data'][0].get('name', 'N/A')}")
                    else:
                        print(f"     ❌ 失败: {data.get('message', '未知错误')}")
                else:
                    print(f"     ❌ HTTP错误: {response.status_code}")
                    
            except Exception as e:
                print(f"     ❌ 异常: {e}")
    
    def test_advanced_search_supplies(self):
        """测试供求高级搜索"""
        print("\n🔍 [测试] 供求高级搜索...")
        test_cases = [
            {
                'name': '关键词搜索',
                'params': {'keyword': '铁观音'}
            },
            {
                'name': '类型筛选',
                'params': {'type': '供应'}
            },
            {
                'name': '状态筛选',
                'params': {'status': 'active'}
            },
            {
                'name': '组合搜索',
                'params': {'keyword': '铁观音', 'type': '供应', 'status': 'active'}
            }
        ]
        
        for test_case in test_cases:
            try:
                print(f"   - 测试: {test_case['name']}")
                response = self.session.get(
                    f"{self.base_url}/api/advanced-search/supplies",
                    params=test_case['params'],
                    timeout=TIMEOUT
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data.get('status') == 'success':
                        print(f"     ✅ 成功: {len(data['data'])} 条结果")
                        if data['data']:
                            print(f"     📋 示例: {data['data'][0].get('title', 'N/A')}")
                    else:
                        print(f"     ❌ 失败: {data.get('message', '未知错误')}")
                else:
                    print(f"     ❌ HTTP错误: {response.status_code}")
                    
            except Exception as e:
                print(f"     ❌ 异常: {e}")
    
    def test_advanced_search_clearance(self):
        """测试清仓高级搜索"""
        print("\n🔍 [测试] 清仓高级搜索...")
        test_cases = [
            {
                'name': '关键词搜索',
                'params': {'keyword': '普洱'}
            },
            {
                'name': '分类筛选',
                'params': {'category': '普洱茶'}
            },
            {
                'name': '价格范围',
                'params': {'price_range': '50-300'}
            },
            {
                'name': '组合搜索',
                'params': {'keyword': '普洱', 'category': '普洱茶', 'price_range': '50-300'}
            }
        ]
        
        for test_case in test_cases:
            try:
                print(f"   - 测试: {test_case['name']}")
                response = self.session.get(
                    f"{self.base_url}/api/advanced-search/clearance",
                    params=test_case['params'],
                    timeout=TIMEOUT
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data.get('status') == 'success':
                        print(f"     ✅ 成功: {len(data['data'])} 条结果")
                        if data['data']:
                            print(f"     📋 示例: {data['data'][0].get('name', 'N/A')}")
                    else:
                        print(f"     ❌ 失败: {data.get('message', '未知错误')}")
                else:
                    print(f"     ❌ HTTP错误: {response.status_code}")
                    
            except Exception as e:
                print(f"     ❌ 异常: {e}")
    
    def test_advanced_search_content(self):
        """测试内容高级搜索"""
        print("\n🔍 [测试] 内容高级搜索...")
        test_cases = [
            {
                'name': '关键词搜索',
                'params': {'keyword': '茶艺'}
            },
            {
                'name': '类型筛选',
                'params': {'type': 'art'}
            },
            {
                'name': '标签筛选',
                'params': {'tag': '冲泡技巧'}
            },
            {
                'name': '组合搜索',
                'params': {'keyword': '茶艺', 'type': 'art', 'tag': '冲泡技巧'}
            }
        ]
        
        for test_case in test_cases:
            try:
                print(f"   - 测试: {test_case['name']}")
                response = self.session.get(
                    f"{self.base_url}/api/advanced-search/content",
                    params=test_case['params'],
                    timeout=TIMEOUT
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data.get('status') == 'success':
                        print(f"     ✅ 成功: {len(data['data'])} 条结果")
                        if data['data']:
                            print(f"     📋 示例: {data['data'][0].get('title', 'N/A')}")
                    else:
                        print(f"     ❌ 失败: {data.get('message', '未知错误')}")
                else:
                    print(f"     ❌ HTTP错误: {response.status_code}")
                    
            except Exception as e:
                print(f"     ❌ 异常: {e}")
    
    def test_edge_cases(self):
        """测试边界情况"""
        print("\n🔬 [测试] 边界情况测试...")
        
        # 空搜索
        try:
            print("   - 测试: 空搜索")
            response = self.session.get(f"{self.base_url}/api/advanced-search/markets", timeout=TIMEOUT)
            if response.status_code == 200:
                data = response.json()
                print(f"     ✅ 空搜索正常: {len(data.get('data', []))} 条结果")
            else:
                print(f"     ❌ 空搜索失败: HTTP {response.status_code}")
        except Exception as e:
            print(f"     ❌ 空搜索异常: {e}")
        
        # 无效价格范围
        try:
            print("   - 测试: 无效价格范围")
            response = self.session.get(
                f"{self.base_url}/api/advanced-search/newarrivals",
                params={'price_range': 'invalid-price'},
                timeout=TIMEOUT
            )
            if response.status_code == 200:
                data = response.json()
                print(f"     ✅ 无效价格范围处理正常: {len(data.get('data', []))} 条结果")
            else:
                print(f"     ❌ 无效价格范围处理失败: HTTP {response.status_code}")
        except Exception as e:
            print(f"     ❌ 无效价格范围处理异常: {e}")
        
        # 无效日期范围
        try:
            print("   - 测试: 无效日期范围")
            response = self.session.get(
                f"{self.base_url}/api/advanced-search/supplies",
                params={'date_range': 'invalid-date'},
                timeout=TIMEOUT
            )
            if response.status_code == 200:
                data = response.json()
                print(f"     ✅ 无效日期范围处理正常: {len(data.get('data', []))} 条结果")
            else:
                print(f"     ❌ 无效日期范围处理失败: HTTP {response.status_code}")
        except Exception as e:
            print(f"     ❌ 无效日期范围处理异常: {e}")
    
    def test_performance(self):
        """性能测试"""
        print("\n⚡ [测试] 性能测试...")
        
        start_time = time.time()
        
        try:
            # 并发搜索测试
            import concurrent.futures
            
            def search_task(search_type, params):
                try:
                    response = self.session.get(
                        f"{self.base_url}/api/advanced-search/{search_type}",
                        params=params,
                        timeout=TIMEOUT
                    )
                    if response.status_code == 200:
                        data = response.json()
                        return f"{search_type}: {len(data.get('data', []))} 条结果"
                    else:
                        return f"{search_type}: HTTP {response.status_code}"
                except Exception as e:
                    return f"{search_type}: 异常 {e}"
            
            search_tasks = [
                ('markets', {'keyword': '茶'}),
                ('newarrivals', {'category': '绿茶'}),
                ('supplies', {'type': '供应'}),
                ('clearance', {'category': '普洱茶'}),
                ('content', {'type': 'recommend'})
            ]
            
            with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
                futures = [executor.submit(search_task, search_type, params) 
                          for search_type, params in search_tasks]
                results = [future.result() for future in concurrent.futures.as_completed(futures)]
            
            end_time = time.time()
            duration = end_time - start_time
            
            print(f"✅ 并发搜索测试完成，耗时: {duration:.2f}秒")
            print(f"   - 平均响应时间: {duration/len(results):.2f}秒")
            
            for result in results:
                print(f"   - {result}")
                
        except Exception as e:
            print(f"❌ 性能测试失败: {e}")
    
    def run_all_tests(self):
        """运行所有测试"""
        print("🚀 [高级搜索测试] 开始全面测试...")
        print("=" * 50)
        
        try:
            # 基础功能测试
            if not self.test_health():
                print("❌ 服务器连接失败，停止测试")
                return
            
            self.test_filter_options()
            self.test_advanced_search_markets()
            self.test_advanced_search_newarrivals()
            self.test_advanced_search_supplies()
            self.test_advanced_search_clearance()
            self.test_advanced_search_content()
            
            # 边界情况测试
            self.test_edge_cases()
            
            # 性能测试
            self.test_performance()
            
            print("\n🎉 [测试完成] 所有测试通过!")
            print("=" * 50)
            
        except Exception as e:
            print(f"\n❌ [测试失败] 测试过程中出现错误: {e}")

def main():
    """主函数"""
    tester = AdvancedSearchTester()
    tester.run_all_tests()

if __name__ == "__main__":
    main() 